tensorflow-keras/src/main/java/org/tensorflow/keras/backend/K.java

```diff
@@ -0,0 +1,713 @@
+   * @param yTrue Optional label `Tensor` whose dimensions match `y_pred`.
+   * @param sampleWeight Optional weight scalar or `Tensor` whose dimensions match `y_pred`.
+   * @return Tuple of `y_pred`, `y_true` and `sample_weight`. Each of them possibly has the last
+   *     dimension squeezed, `sample_weight` could be extended by one dimension. If `sample_weight`
+   *     is null, (y_pred, y_true) is returned.
+   */
+  public static Tuple squeezeOrExpandDimensions(
+      Ops tf, Operand yTrue, Operand yPred, Operand sampleWeight) {
+    Tuple tuple = new Tuple(yTrue, yPred);
+    Shape ypredShape = yPred.asOutput().shape();
+    long ypredRank = ypredShape.numDimensions();
+
+    if (yTrue != null) {
+      Shape ytrueShape = yTrue.asOutput().shape();
+      long ytrueRank = ytrueShape.numDimensions();
+      if (ytrueRank != Shape.UNKNOWN_SIZE && ypredRank != Shape.UNKNOWN_SIZE) {
+        // Use static rank for `y_true` and `y_pred`.
+        if (ypredRank - ytrueRank != 1 || ypredShape.size(-1) == 1) {
+          // y_true, y_pred = confusion_matrix.remove_squeezable_dimensions(y_true, y_pred)
+          tuple = ConfusionMatrix.removeSqueezableDimensions(tf, yTrue, yPred);
+        }
+      } else { // use dynamic rank
+        tuple = ConfusionMatrix.removeSqueezableDimensions(tf, yTrue, yPred);
+      }
+    }
+    if (sampleWeight == null) {
+      return tuple;
+    }
+    Shape weightsShape = sampleWeight.asOutput().shape();
+    long weightsRank = weightsShape.numDimensions();
+    if (weightsRank == 0) { // scalar
+      return new Tuple(yTrue, yPred, sampleWeight);
+    }
+
+    if (ypredRank != Shape.UNKNOWN_SIZE && weightsRank != Shape.UNKNOWN_SIZE) {
+
+      if (weightsRank - ypredRank == 1) {
+        sampleWeight = tf.squeeze(sampleWeight);
+      } else if (ypredRank - weightsRank == 1) {
+        sampleWeight = tf.expandDims(sampleWeight, tf.constant(-1L));
+      }
+      return new Tuple(yTrue, yPred, sampleWeight);
+    }
+    // Use dynamic rank.
+    Operand weightsRankTensor = tf.rank(sampleWeight);
+    Operand rankDiff = tf.math.sub(weightsRankTensor, tf.rank(yPred));
+    sampleWeight =
+        tf.select(
+            tf.math.equal(weightsRankTensor, tf.constant(0)),
+            sampleWeight,
+            maybeAdjustWeights(tf, sampleWeight, rankDiff));
+    return new Tuple(yTrue, yPred, sampleWeight);
+  }
+
+  private static Operand maybeAdjustWeights(Ops tf, Operand sampleWeight, Operand rankDiff) {
+    return tf.select(
+        tf.math.equal(rankDiff, tf.constant(1)),
+        tf.squeeze(sampleWeight, Squeeze.axis(Arrays.asList(-1L))),
+        maybeExpandWeights(tf, sampleWeight, rankDiff));
+  }
+
+  private static Operand maybeExpandWeights(Ops tf, Operand sampleWeight, Operand rankDiff) {
+    return tf.select(
+        tf.math.equal(rankDiff, tf.constant(-1)),
+        tf.expandDims(sampleWeight, tf.constant(-1)),
+        sampleWeight);
+  }
+}
```

Comment:
At some point, we will need to document all these methods but I guess it is fine doing it later when the interface is more stable

Reply:
I have added JavaDoc to my copy, that I haven't pushed yet.

As far as `NdArraySequence`, could be have a common method on each numeric type that returns` java.lang.Number`?

Reply:
I think that would improve the experience at the user level (while passing the burden to the author :) ). Most of the `NdArray` library has been written with that spirit.

With type erasure though, you will need to provide a distinct name for each variant, i.e. something like:

`public NdArraySequence<FloatNdArray> getTensorFloats(Ops tf, Operand<TFloat32> operand)`